### PR TITLE
Fix performance degradation in loss computation

### LINKED ISSUE #689: Performance bug in loss computation
### What happened?

The performance of loss computation degraded substantially with #533. E.g. with ERA5, forecast 2-steps: loss computation takes 0.53 sec while forward pass is 0.36 sec.

### What are the steps to reproduce the bug?

Train any model and time loss computation.

### Version

develop

### Platform (OS and architecture)

ALL

### Relevant log output

```shell

```

### Accompanying data

_No response_

### Organisation

_No response_

### PR DESCRIPTION
**Pull request by clessig**
## Description

Loss value before/after  : 0.9835155606269836 / 0.9835155606269836
 
ERA5, forecast 2-steps:
Time for loss computation before/after:  0.5396285057067871 / 0.00821542739868164 
(in relation, forward pass is 0.36)
 
ERA5 + SEVIRI, forecast 2-steps:
Time for loss computation before/after: 0.19628429412841797 / 0.04020524024963379

## Type of Change

-   [X] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update

## Issue Number

Closes #689 

## Code Compatibility

-   [X] I have performed a self-review of my code

### Code Performance and Testing

-   [X] I ran the `uv run train` and (if necessary) `uv run evaluate` on a least one GPU node and it works 
-   [ ] If the new feature introduces modifications at the config level, I have made sure to have notified the other software developers through Mattermost and updated the paths in the `$WEATHER_GENERATOR_PRIVATE` directory

### Dependencies

-   [ ] I have ensured that the code is still pip-installable after the changes and runs
-   [ ] I have tested that new dependencies themselves are pip-installable.
-   [ ] I have not introduced new dependencies in the inference portion of the pipeline

### Documentation

-   [ ] My code follows the style guidelines of this project
-   [ ] I have updated the documentation and docstrings to reflect the changes
-   [ ] I have added comments to my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/weathergen/train/loss.py [hunk 1]
```diff
@@ -91,3 +82,62 @@ def kernel_crps(target, ens, mu, stddev, fair=True):
...
+    pred: torch.Tensor,
+    weights_channels: torch.Tensor | None,
+    weights_points: torch.Tensor | None,
+):
+    """
+    Compute weighted MSE loss for one window or step
+
+    The function implements:
+
+    loss = Mean_{channels}( weight_channels * Mean_{data_pts}( (target - pred) * weights_channels ))
```
Comment (MatKbauer) [inline src/weathergen/train/loss.py]: Should it be `weights_points` instead of `weights_channels` in the right-most multiplication?

File: src/weathergen/train/loss.py [hunk 2]
```diff
@@ -91,3 +82,62 @@ def kernel_crps(target, ens, mu, stddev, fair=True):
...
+    1. weight the rows of (target - pred) by wp = weights_points
+    2. take the mean over the row
+    3. weight the collapsed cols by wc = weights_channels
+    4. take the mean over the channel-weighted cols
+
+    Params:
+        target : shape ( num_data_points , num_channels )
+        target : shape ( ens_dim , num_data_points , num_channels)
+        weights_channels : shape = (num_channels,)
+        weights_points : shape = (num_data_points)
```
Comment (MatKbauer) [inline src/weathergen/train/loss.py:128]: Is `weights_points` a weighting for each individual encoded data point on the HEALPix? Can we apply latitude-weighting here? Maybe rename to `weights_latlon` or `weights_spatial` if so?

File: src/weathergen/train/loss.py [hunk 3]
```diff
@@ -91,3 +82,62 @@ def kernel_crps(target, ens, mu, stddev, fair=True):
...
+        |                      |    |  |
+        |     target - pred    | x  |wp|
+        |                      |    |  |
+        |                      |    |  |
+        |                      |    |  |
+        ------------------------     -
+                    x
+        ------------------------
+        |          wc          |
+        ------------------------
```
Comment (MatKbauer) [inline src/weathergen/train/loss.py:114]: Nice!

File: src/weathergen/train/loss_calculator.py [hunk 4]
```diff
@@ -77,145 +77,101 @@ def __init__(
...
-                # for detailed logging
-                losses_all[strm.name][i_ch, i_lfct] += loss_ch.item()
+            # accumulate loss
+            loss_lfct = loss_lfct + loss
+            losses_chs += loss_chs.detach()
+            ctr_substeps += 1 if loss > 0.0 else 0
 
             # normalize over forecast steps in window
-            losses_all[strm.name][i_ch, i_lfct] /= ctr_substeps if ctr_substeps > 0 else 0.0
+            losses_chs /= ctr_substeps if ctr_substeps > 0 else 0.0
```
Comment (MatKbauer) [inline src/weathergen/train/loss_calculator.py]: Should this line be unindendet one tab? I'm wondering why we normalize `losses_chs` with `ctr_substeps`, while `ctr_substeps` is incremented in this loop. This leads to multiple normalizations based on different incremented values of `ctr_substeps`. Also, `loss_chs` is devided by zero when `ctr_substeps == 0`.

File: src/weathergen/train/loss_calculator.py [hunk 5]
```diff
@@ -77,145 +77,101 @@ def __init__(
...
-                           which influences mask construction.
+        # Determine stream and channel loss weights based on the current stage
+        if self.stage == TRAIN:
+            # set loss_weights to 1. when not specified
+            stream_info_loss_weight = stream_info.get("loss_weight", 1.0)
+            weights_channels = (
+                torch.tensor(stream_info["channel_weight"]).to(device=device, non_blocking=True)
+                if "channel_weight" in stream_info
+                else None
+            )
```
Comment (MatKbauer) [inline src/weathergen/train/loss_calculator.py:100]: Is this `.to(device=device, non_blocking=True)` notation robust to mixed-precision scenarios? If `type(weights_channels) != type(unweighted_loss)`, this might cause problems.

File: src/weathergen/train/loss.py (hunk 2 above)
Comment (clessig) [reply src/weathergen/train/loss.py:128]: Points can be space or space-time, depending on tok_timestep

File: src/weathergen/train/loss_calculator.py (hunk 4 above)
Comment (clessig) [reply src/weathergen/train/loss_calculator.py]: Yes!

**Review by MatKbauer — APPROVED** [review]
Nice catch with the assertion. Looks good.
